Userland/Libraries/LibWeb/Painting/StackingContext.cpp

```diff
@@ -146,61 +154,55 @@ void StackingContext::paint_internal(PaintContext& context) const
     // NOTE: Non-positioned descendants that establish stacking contexts with z-index `0` or `auto` are also painted here.
     // FIXME: There's more to this step that we have yet to understand and implement.
-    m_box.paint_box()->for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
-        auto const& z_index = paint_box.computed_values().z_index();
+    paintable().for_each_in_subtree_of_type<PaintableBox>([&](PaintableBox const& paint_box) {
+        auto z_index = paint_box.computed_values().z_index();
         if (auto* child = paint_box.stacking_context()) {
-            if (!z_index.has_value() || z_index.value() == 0)
+            if (z_index.value_or(0) == 0)
                 paint_child(child);
             return TraversalDecision::SkipChildrenAndContinue;
-        }
-        if (z_index.has_value() && z_index.value() != 0)
-            return TraversalDecision::Continue;
-        if (!paint_box.layout_box().is_positioned())
+        } else {
+            if (z_index.value_or(0) == 0 && paint_box.layout_box().is_positioned()) {
+                paint_box_background(paint_box.layout_box());
+                paint_box_content(paint_box.layout_box());
+                paint_box_overlay(paint_box.layout_box());
+            }
             return TraversalDecision::Continue;
-        // At this point, `paint_box` is a positioned descendant with z-index: auto
-        // but no stacking context of its own.
-        // FIXME: This is basically duplicating logic found elsewhere in this same function. Find a way to make this more elegant.
-        paint_node(paint_box.layout_box(), context, PaintPhase::Background);
-        paint_node(paint_box.layout_box(), context, PaintPhase::Border);
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::BackgroundAndBorders);
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::Floats);
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::BackgroundAndBordersForInlineLevelAndReplaced);
-        paint_node(paint_box.layout_box(), context, PaintPhase::Foreground);
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::Foreground);
-        paint_node(paint_box.layout_box(), context, PaintPhase::FocusOutline);
-        paint_node(paint_box.layout_box(), context, PaintPhase::Overlay);
-        paint_descendants(context, paint_box.layout_box(), StackingContextPaintPhase::FocusAndOverlay);
-
-        return TraversalDecision::Continue;
+        }
     });
 
     // Draw other positioned descendants (step 9)
     for (auto* child : m_children) {
-        if (child->m_box.computed_values().z_index().has_value() && child->m_box.computed_values().z_index().value() >= 1)
+        if (child->m_box.computed_values().z_index().has_value() && child->m_box.computed_values().z_index().value() > 0)
```

Comment:
The spec says: `Stacking contexts formed by positioned descendants with z-indices greater than or equal to 1 in z-index order (smallest first) then tree order.`

We try to match the spec language as close as possible so I think we should keep the `>= 1`

(Unless I found an old spec)

Reply:
hmm, yes, well, when reading the code you basically have three classes: negative, 0/auto, positive, and I found it much clearer this way that this is the case. i find the spec wording unnneccessarily confusing when it clearly sorts things into three buckets, but uses different formulations for the buckets. but again, not stron about it

Reply:
but I do see your point..changed it back :)